ksmt-core/src/main/kotlin/org/ksmt/expr/Array.kt

```diff
@@ -31,6 +35,129 @@ sealed class KArrayStoreBase<A : KArraySortBase<R>, R : KSort>(
             addAll(this@KArrayStoreBase.indices)
             add(value)
         }.uncheckedCast()
+
+    private var type: ArrayStoreType = ArrayStoreType.NOT_INITIALIZED
+    private var interpretedStoresAmount: Int = 0
+    private var interpretedStores: PersistentMap<Any, KArrayStoreBase<A, R>>? = null
+    private var nextUninterpretedArray: KExpr<A>? = null
+
+    abstract fun indicesAreInterpreted(): Boolean
+
+    /**
+     * Find an array expression containing the value for the provided indices.
+     *
+     * Returns:
+     * 1. Indices match the current store expression indices -> current store.
+     * 2. We can't determine whether indices match or not -> current store.
+     * 3. Indices definitely don't match current store indices
+     * (e.g. both are interpreted values) -> apply this operation on a nested array.
+     * */
+    abstract fun findArrayToSelectFrom(indices: List<KExpr<*>>): KExpr<A>
+
+    /**
+     * Analyze store to provide faster index lookups when
+     * store indices are interpreted values.
+     * */
+    fun analyzeStore() {
+        // Store was analyzed already. Nothing to do here.
+        if (type != ArrayStoreType.NOT_INITIALIZED) return
+
+        // Store indices are not interpreted values. We can't optimize lookups.
+        if (!indicesAreInterpreted()) {
+            type = ArrayStoreType.UNINTERPRETED
+            return
+        }
+
+        type = ArrayStoreType.INTERPRETED
+
+        /**
+         * Current store is the first store with interpreted indices.
+         * We don't need to initialize map.
+         * */
+        if (array !is KArrayStoreBase<A, *> || array.type != ArrayStoreType.INTERPRETED) {
```

Comment:
I didn't get it. Why don't we need to initialize map if we have an KArrayStoreBase instance here? How do we check that there was no interpreted array earlier? Just because we won't go so deep if we already encountered an interpreted indices?

Anyway, please, add more clarifying comments here